PD-84 Update Replication Tutorials (#1839)
truenas/documentation: content/SCALE/SCALETutorials/DataProtection/Replication/ReplicationWithEncryptionSCALE.md

---
title: "Setting Up a Encrypted Replication Task"
description: "This article provides instructions on adding a replication task to a remote system and using encryption."
weight: 40
aliases:
tags:
- scalereplication
- scalebackup
---

{{< toc >}}

## Using Encryption in Replication Tasks

> TrueNAS SCALE replication allows users to create replicated snapshots of data stored in encrypted pools, datasets or zvols that on their SCALE system as a way to back up stored data to a remote system. You can use encrypted datasets in a local replication.

Reviewer: ```suggestion
TrueNAS SCALE replication allows users to create replicated snapshots of data stored in encrypted pools, datasets, or zvols on their SCALE system to back up stored data to a remote system. You can use encrypted datasets in a local replication.
```

{{< hint warning >}}

> You can set up a replication task for a dataset encrypted with a passphrase or a hex encryption key, but you must unlock the dataset before the task runs or the task fails.

Reviewer: ```suggestion
You can set up a replication task for a dataset encrypted with a passphrase or a hex encryption key, but you must unlock the dataset before running the task, or it fails.
```

{{< /hint>}}

> With the implementation of rootless login and the admin user, when setting up remote replication tasks when logged in as an admin user requires selecting **Use Sudo For ZFS Commands**. 

Reviewer: ```suggestion
With the implementation of rootless login and the admin user, setting up remote replication tasks when logged in as an admin user requires selecting **Use Sudo For ZFS Commands**. 
```

{{< include file="/content/_includes/ReplicationIntroSCALE.md" type="page" >}}

> Remote replication with datasets also require an SSH connection in TrueNAS. You can use an existing SSH connection if it has the same user credentials you want to use for the new replication task. 

Reviewer: ```suggestion
Remote replication with datasets also requires an SSH connection in TrueNAS. You can use an existing SSH connection if it has the same user credentials you want to use for the new replication task. 
```

## Setting Up a Simple Replication Task Overview 

> This section provides a simple overview of setting up a remote replication task for an encrypted dataset. 
> It also covers the related steps you should take prior to configuring the replication task. 

Reviewer: ```suggestion
This section provides a simple overview of setting up a remote replication task for an encrypted dataset. 
It also covers the steps you should take before configuring the replication task. 
```

1. Set up the data storage for where you want to save replicated snapshots. 
   
2. Make sure the admin user has a home directory assigned. 
   In the SCALE Bluefin early release, when creating the admin user at installation the home directory default is set to **/nonexistent**. To create an SSH connection to use in a remote replication you must assign a home directory path.

   Later releases of SCALE Bluefin set the admin user home directory to one created by SCALE during the installation process, but you need to select the option to create the admin user home directory.

3. Create an SSH connection between the local SCALE system and the remote system. 
   You can do this from either **Credentials > Backup Credentials > SSH Connection** and clicking **Add** or from the **Replication Task Wizard** using the **Generate New** option in the settings for the remote system.

4. Unlock the encrypted dataset(s) and export the encryption key to a text editor like Notepad.

5. Go to **Data Protection > Replication Tasks** and click **Add** to open the **Replication Task Wizard**. 
   You then specify the from and to sources, task name, and set the schedule.
  

>    Setting options change based on the source selections. Replicating to or from a local source does not requires an SSH connection.

Reviewer 2: Suggest reword: 'Replicating to or from a local source..." to "A to or from source that is local, does not..." the first one almost sounds like any time you are replicating to or from a local source (which is always) you never need an SSH connection. Or perhaps "When a source (to or from) is remote, an SSH Connection is required"

   
This completes the general process for all replication tasks.
{{< /expand >}}

## Creating a Remote Replication Task for an Encrypted Dataset

> To streamline creating simple replication tasks use the **Replication Task Wizard** to create and copy ZFS snapshots to another system. 
> The wizard assists with creating a new SSH connection and automatically creates a periodic snapshot task for sources that have no existing snapshots.

Reviewer: ```suggestion
To streamline creating simple replication tasks, use the **Replication Task Wizard** to create and copy ZFS snapshots to another system. 
The wizard assists with creating a new SSH connection and automatically creates a periodic snapshot task for sources with no existing snapshots.
```

> If you have an existing replication task, you can select it on the **Load Previous Replication Task** dropdown list to load the configuration settings for that task into the wizard, and then make change such as assigning it a different destination, select encryption options, schedule, or retention lifetime, etc. 
> Saving changes to the configuration creates a new replication task without altering the task you loaded into the wizard.
> This saves some time when creating multiple replication tasks between the same two systems.

Reviewer: ```suggestion
If you have an existing replication task, you can select it on the **Load Previous Replication Task** dropdown list to load the configuration settings for that task into the wizard, then make changes such as assigning it a different destination, selecting encryption options, schedule, retention lifetime, etc. 
Saving changes to the configuration creates a new replication task without altering the task you loaded into the wizard.
This saves some time when creating multiple replication tasks between the same two systems.
```

{{< include file="/content/_includes/ReplicationCreateDatasetAndAdminHomeDirSteps.md" type="page" >}}

> 3. Unlock the source dataset and export the encryption key to a text editor such as Notepad.
>    Go to **Datasets** select the source dataset, locate the **ZFS Encryption** widget and unlock the dataset if locked. 
>    Export the key and paste it in any text editor such as Notepad. If you set up encryption to use a passphrase, you do not need to export a key.

Reviewer: ```suggestion
3. Unlock the source dataset and export the encryption key to a text editor like Notepad.
   Go to **Datasets**, select the source dataset, locate the **ZFS Encryption** widget, and unlock the dataset if locked. 
   Export the key, then paste it into any text editor, like Notepad. You do not need to export a key if you set up encryption to use a passphrase.
```

4. Go to **Data Protection** and click **Add** on the **Replication Tasks** widget to open the **Replication Task Wizard**. Configure the following settings:
   
   ![CreateRemoteReplicationTask](/images/SCALE/22.12/CreateRemoteReplicationTask.png "New Remote Replication Task")
   
   a. Select either **On this System** or **On a Different System** on the **Source Location** dropdown list. 
      If your source is a remote system, select **On a Different System**. The **Destination Location** automatically changes to **On this System**.       
      If your source is the local TrueNAS SCALE system, you must select **On a Different System** from the **Destination Location** dropdown list to do remote replication. 
      

>       TrueNAS shows the number snapshots available for replication.

Reviewer 2: missing word: 'number **of** snapshots'

>     b. Select an existing SSH connection to the remote system.

Reviewer 2: should this be an 'or'? or reword to suggest you do one or the other. currently it looks like you do both.

       Select **Create New** to open the **[New SSH Connection](#configure-a-new-ssh-connection)** configuration screen.
    
    c. Browse to the source pool/dataset(s), then click on the dataset(s) to populate the **Source** with the path. 

>        You can select multiple sources or manually type the names into the **Source** field. 

Reviewer 2: might want to add after 'manually type.. field.' the following '(separated by commas)'

       If not selected you need to issue the cli `zfs allow` command in Shell on the remote system. 

    g. Select **Replicate Custom Snapshots**, then accept the default value in **Naming Schema**. 
       Remote sources require entering a snapshot naming schema to identify the snapshots to replicate. 
       A naming schema is a pattern of naming custom snapshots you want to replicate. 
       If you want to change the default schema, enter the name and [strftime(3)](https://man7.org/linux/man-pages/man3/strftime.3.html) %Y, %m, %d, %H, and %M strings that match the snapshots to include in the replication. 
       Separate entries by pressing <kbd>Enter</kbd>. The number of snapshots matching the patterns display.
    
    ![RemoteReplicateSnapshotAndNameSchema](/images/SCALE/22.12/RemoteReplicateSnapshotAndNameSchema.png "Replicate Custom Snapshot and Naming Schema")

   h. (Optional) Enter a name for the snapshot in **Task Name**. 
      SCALE populates this field with the default name using the source and destination paths separated by a hyphen, but this default can make locating the snapshot in destination dataset a challenge. 

>       To make it easier to find the snapshot, give it name easy for you to identify. For example, a replicated task named *dailyfull* for a full file system snapshot taken daily. 

Reviewer 2: Missing a word: 'give it name' to 'give it a name'
Also, 'name easy for you to identify' to 'name for you to easily identify'
